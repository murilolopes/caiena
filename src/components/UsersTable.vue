<template>
  <div>
    <b-table
      v-if="$store.state.users"
      show-empty
      :fields="fields"
      :items="$store.state.users"
      :busy.sync="$store.state.busy"
      empty-html="No results found"
    >
      <template #table-busy>
        <div class="text-center text-danger my-2">
          <b-spinner class="align-middle"></b-spinner>
          <strong>Loading...</strong>
        </div>
      </template>
      <template #cell(avatar_url)="item">
        <img :src="item.item.avatar_url" width="50" height="50" />
      </template>
      <template #cell(html_url)="item">
        <a :href="item.item.html_url" target="_blank">Link</a>
      </template>
    </b-table>
    <b-pagination
      v-if="$store.state.users"
      v-model="$store.state.selectedPage"
      :total-rows="$store.state.total_count"
      :per-page="$store.state.per_page"
      @change="changePage"
    />
  </div>
</template>

<script>
export default {
  name: "UsersTable",
  data() {
    return {
      fields: [
        { key: "avatar_url", sortable: true },
        { key: "login", sortable: true },
        { key: "html_url", sortable: true },
      ],
    };
  },
  methods: {
    changePage(page) {
      this.$store.dispatch("changePage", page);
    },
  },
};
</script>

<style></style>
