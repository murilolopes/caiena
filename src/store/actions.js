import GithubService from "@/services/github";
import Swal from "sweetalert2";

export default {
  async search({ commit, dispatch, rootState }, term = null) {
    return new Promise((resolve, reject) => {
      commit("SET_BUSY");

      if (term) {
        commit("SET_SEARCH_TERM", term);
        commit("SET_SELECTED_PAGE", 1);
      }

      const payload = {
        q: rootState.term,
        page: rootState.selected_page,
        per_page: rootState.per_page,
      };

      GithubService.fetchUsers(payload)
        .then((result) => {
          commit("SET_USERS", result.items);
          commit("SET_TOTAL_COUNT", result.total_count);
          resolve(result);
        })
        .catch((error) => {
          dispatch("clearResults");
          Swal.fire({
            title: "Error!",
            text: error.message,
            icon: "error",
          });
          commit("SET_ERRORS", error);
          reject(error);
        })
        .finally(() => {
          commit("SET_BUSY");
        });
    });
  },

  clearResults({ commit }) {
    commit("SET_SEARCH_TERM", "");
    commit("SET_SELECTED_PAGE", 1);
    commit("SET_USERS", []);
    commit("SET_TOTAL_COUNT", 0);
  },

  changePage({ commit, dispatch }, page) {
    commit("SET_SELECTED_PAGE", page);
    dispatch("search");
  },
};
