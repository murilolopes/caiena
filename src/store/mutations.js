export default {
  SET_USERS(state, payload) {
    state.users = payload;
  },
  SET_SEARCH_TERM(state, payload) {
    state.term = payload;
  },
  SET_TOTAL_COUT(state, payload) {
    if (payload > state.per_page) state.total_count = payload - state.per_page;
    if (payload <= state.per_page) state.total_count = payload;
  },
  SET_ERRORS(state, payload) {
    state.errors = payload;
  },
  SET_SELECTED_PAGE(state, payload) {
    state.selected_page = payload;
  },
  SET_BUSY(state) {
    state.busy = !state.busy;
  },
};
