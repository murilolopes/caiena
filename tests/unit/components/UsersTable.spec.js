import Vue from "vue";
import Vuex from "vuex";
import BootstrapVue from "bootstrap-vue";
import { BTable, BPagination } from "bootstrap-vue";
import actions from "../../../src/store/actions";
import mutations from "../../../src/store/mutations";
import { mount, createLocalVue } from "@vue/test-utils";
import UsersTable from "./../../../src/components/UsersTable.vue";

Vue.use(BootstrapVue);
const localVue = createLocalVue();
localVue.use(Vuex);

describe("UsersTable.vue", () => {
  test("table and pagination should be visible only when state.users has value", () => {
    const store = new Vuex.Store({
      state: { users: [{ key: "value" }] },
    });
    const wrapper = mount(UsersTable, { store, localVue });
    const table = wrapper.findComponent(BTable);
    const pagination = wrapper.findComponent(BPagination);

    expect(table.isVisible()).toBeTruthy();
    expect(pagination.isVisible()).toBeTruthy();
  });

  test("changePage method should dispatch changePage action with page as a parameter", () => {
    const store = new Vuex.Store({
      actions,
      mutations,
    });
    jest.spyOn(store, "dispatch");
    const wrapper = mount(UsersTable, { store, localVue });
    wrapper.vm.changePage(1);

    expect(store.dispatch).toHaveBeenCalledWith("changePage", 1);
  });
});
