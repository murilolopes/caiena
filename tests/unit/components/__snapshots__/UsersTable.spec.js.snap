// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTable.vue component should be rendered correctly 1`] = `
<div>
  <table
    aria-busy="false"
    aria-colcount="3"
    class="table b-table"
    role="table"
  >
    <!---->
    <!---->
    <thead
      class=""
      role="rowgroup"
    >
      <!---->
      <tr
        class=""
        role="row"
      >
        <th
          aria-colindex="1"
          aria-sort="none"
          class=""
          role="columnheader"
          scope="col"
          tabindex="0"
        >
          <div>
            Avatar Url
          </div>
          <span
            class="sr-only"
          >
             (Click to sort Ascending)
          </span>
        </th>
        <th
          aria-colindex="2"
          aria-sort="none"
          class=""
          role="columnheader"
          scope="col"
          tabindex="0"
        >
          <div>
            Login
          </div>
          <span
            class="sr-only"
          >
             (Click to sort Ascending)
          </span>
        </th>
        <th
          aria-colindex="3"
          aria-sort="none"
          class=""
          role="columnheader"
          scope="col"
          tabindex="0"
        >
          <div>
            Html Url
          </div>
          <span
            class="sr-only"
          >
             (Click to sort Ascending)
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <!---->
      <tr
        class=""
        role="row"
      >
        <td
          aria-colindex="1"
          class=""
          role="cell"
        >
          <img
            height="50"
            width="50"
          />
        </td>
        <td
          aria-colindex="2"
          class=""
          role="cell"
        />
        <td
          aria-colindex="3"
          class=""
          role="cell"
        >
          <a
            target="_blank"
          >
            Link
          </a>
        </td>
      </tr>
      <!---->
      <!---->
    </tbody>
    <!---->
  </table>
   
  <ul
    aria-disabled="false"
    aria-label="Pagination"
    class="pagination b-pagination"
    role="menubar"
  >
    <li
      aria-hidden="true"
      class="page-item disabled"
      role="presentation"
    >
      <span
        aria-disabled="true"
        aria-label="Go to first page"
        class="page-link"
        role="menuitem"
      >
        «
      </span>
    </li>
    <li
      aria-hidden="true"
      class="page-item disabled"
      role="presentation"
    >
      <span
        aria-disabled="true"
        aria-label="Go to previous page"
        class="page-link"
        role="menuitem"
      >
        ‹
      </span>
    </li>
    <!---->
    <!---->
    <li
      class="page-item"
      role="presentation"
    >
      <button
        aria-checked="false"
        aria-label="Go to page 1"
        aria-posinset="1"
        aria-setsize="1"
        class="page-link"
        role="menuitemradio"
        tabindex="0"
        type="button"
      >
        1
      </button>
    </li>
    <!---->
    <!---->
    <li
      aria-hidden="true"
      class="page-item disabled"
      role="presentation"
    >
      <span
        aria-disabled="true"
        aria-label="Go to next page"
        class="page-link"
        role="menuitem"
      >
        ›
      </span>
    </li>
    <li
      aria-hidden="true"
      class="page-item disabled"
      role="presentation"
    >
      <span
        aria-disabled="true"
        aria-label="Go to last page"
        class="page-link"
        role="menuitem"
      >
        »
      </span>
    </li>
  </ul>
</div>
`;
